ApiKeyValueMapUpload Command (#3407)
SAP/jenkins-library: documentation/docs/steps/apiKeyValueMapUpload.md

# ${docGenStepName}

## ${docGenDescription}

## Prerequisites

## ${docGenParameters}

## ${docGenConfiguration}

## ${docJenkinsPluginDependencies}

> ## Example

Reviewer: What happens if the target already exists? How does a continuous process prevent to re-create a key?

Author: the same document template is used in all other CPI commands which was merged previously as suggested by Marcus/Oliver

Reviewer: Missing documentation hinders the usage as well as the review. As a consumer, I would like to get as much information as possible. To reuse a template is fine but does not prevent to add step specific information.

Author: fixed

# ${docGenStepName}

## ${docGenDescription}

With this step you can store one or more key value pairs of data stored in a group called a map or key value map.

To consume the ApiKeyValueMapUpload step, proceed as follows:

> * Copy the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions-->service API Management,API portal, which was created under apiportal-apiaccess plan.

Reviewer 2: ```suggestion
* Copy the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions-->service API Management, API portal, which was created under apiportal-apiaccess plan.
```

Reviewer 2: I'm not sure the `-->` will look good in the rendered document. We should build it again and see.

> * Copy the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions-->service API Management, API portal, which was created under apiportal-apiaccess plan.

Reviewer 2: [Use a proper markdown arrow (&rarr;)](https://reimbar.org/dev/arrows/):
```suggestion
* Copy the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions &rarr; service API Management, API portal, which was created under apiportal-apiaccess plan.
```

# ${docGenStepName}

## ${docGenDescription}

* With this step you can store one or more key value pairs of data stored in a group called a map or key value map.

> * To consume the ApiKeyValueMapUpload step, proceed as follows:

Reviewer 2: What about @LauraVeinberga 's [comment](https://github.com/SAP/jenkins-library/pull/3407/#discussion_r792590818) on not having bullet points here.

Author: fixed it

> * A typical flow to consume ApiKeyValueMapUpload step involves:

Author: done

>   * Get the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions for API management instance created under API plan.

Author: fixed

>   * Get the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions for API management instance created under API plan.

Reviewer 2: What was fixed exactly? Looks the same to me.

>   * Get the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions for API management instance created under API plan.

Author: this "API Management, API portal under apiportal-apiaccess plan"  should be fine as we seen in BTP subaccount cockpit under subscriptions

> * Get the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions for service API Management, API portal under apiportal-apiaccess plan.

Reviewer 2: I don't understand this sentence, please rephrase.

Author: done

>   * Store SAP API management service key in the jenkins server as secret text.

Author: done

>   * Store SAP API management service key in the jenkins server as secret text.

Reviewer 2: done

> * Store your service key created for SAP API Management in the the Jenkins server as a secret text.

Reviewer 2: ```suggestion
* Store your service key created for SAP API Management in the Jenkins server as a secret text.
```

Author: done

>   * Create a new jenkins pipeline, which consumes the ApiKeyValueMapUpload step.

Author: done

* Create a new Jenkins pipeline designated for the ApiKeyValueMapUpload step.

> * Execute the pipeline and validate step exection results as explained in the blog [Integration Suite Piper commands](https://blogs.sap.com/2022/01/05/orking-with-integration-suite-piper-commands/)

Reviewer 2: ```suggestion
* Execute the pipeline and validate the step execution results as explained in the blog [Integration Suite Piper commands](https://blogs.sap.com/2022/01/05/orking-with-integration-suite-piper-commands/)
```

Author: done

> 

Reviewer 2: Why the new line here, does that do anything to the resulting documentation, and if so: is that intended?

Author: removed

> * Using ApiKeyValueMapUpload step:

Reviewer 2: has been done

> * Using ApiKeyValueMapUpload step, you can create new API key value map in the API portal.

Reviewer 2: ```suggestion
* Using the ApiKeyValueMapUpload step, you can create a new API key value map in the API portal.
```

Author: done

>   * If API key value map already exist, then command execution would fail.

Reviewer 3: * prevent command execution in case the/an API key value map already exists

Author: done

Reviewer 2: I'm not sure the intention was to add another bullet point into the list. @LauraVeinberga could you elaborate your comment?

> * The ApiKeyValueMapUpload step allows you to prevent command execution in case the/an API key value map already exists.

Reviewer 2: ```suggestion
* The ApiKeyValueMapUpload step allows you to prevent command execution in case an API key value map already exists.
```

Author: done

>   * If API key value map already exist, then delete it and execute the piper step again, which create new API Key value Map.

Reviewer 3: * overwrite an already existing API key value map 

(is this what is meant by this step?)

Author: this delete manually need to be done, no overwrite possible

> * If API key value map already exists, then delete it and execute the piper step again, which create new API Key value Map.

Reviewer 2: ```suggestion
* If API key value map already exists, then delete it and execute the piper step again, which will create a new API Key value Map.
```

Author: done

> * ApiKeyValueMapUpload only supports create operation.

Reviewer 2: So you did not want to reference the other steps that can do the other operations? Is it because the other operations are not supported at all yet in Piper?

Author: yes correct

>   * ApiKeyValueMapUpload only support create operation, but not delete, get, update, which are supported in different piper steps.

Reviewer 3: Should this be in the same bullet list or is it a side note? This is not very clear to me.
Note: The `ApiKeyValueMapUpload ` step only supports the **create** operation. The operations **delete**, **get**, and **update** are not supported in this project "Piper" step.

Author: this gives limitation of this command, so it should be fine i feel.

## Prerequisites

## ${docGenParameters}

## ${docGenConfiguration}

## ${docJenkinsPluginDependencies}

## Example

> Example configuration for the use in a `Jenkinsfile`.

Reviewer 3: Configuration example for a `Jenkinsfile`:

Author: same approach is followed in other piper commands

Author: done

Reviewer 2: > same approach is followed in other piper commands

We need to fix documentation in the other steps as well once we have a better standard, which we are establishing now with this step.

```groovy
apiKeyValueMapUpload script: this
```

> Example for the use in a YAML configuration file (such as `.pipeline/config.yaml`).

Reviewer 3: Configuration example for a YAML file (for example, `.pipeline/config.yaml`):

Author: same approach is followed in other piper commands

Author: done

# ${docGenStepName}

## ${docGenDescription}

> With this step you can store one or more key value pairs of data stored in a group called a map or key value map.<br />

Reviewer 2: Add empty line below, instead of `<br />`
```suggestion
With this step you can store one or more key value pairs of data stored in a group called a map or key value map.

```

To consume the ApiKeyValueMapUpload step, proceed as follows:

* Get the SAP API management service key from the SAP BTP sub account cockpit, under instance and subscriptions for service API Management, API portal under API plan.
* Store your service key created for SAP API Management in the the Jenkins server as a secret text.
* Create a new Jenkins pipeline designated for the ApiKeyValueMapUpload step.
* Execute the pipeline and validate step exection results as explained in the blog [Integration Suite Piper commands](https://blogs.sap.com/2022/01/05/orking-with-integration-suite-piper-commands/)

> The ApiKeyValueMapUpload step allows you to:

Reviewer 2: Add an empty line above

Author: done

* You can create new API key value map in the API portal.

> * Prevent command execution in case the/an API key value map already exists.

Reviewer 2: Is this true?

Author: can you shared generated document to see how it looks?

Reviewer 2: @rodibrin Could you build and share the latest changes?

Reviewer 2: > Please find one writing style. This bullet point is a half sentence that does not make sense without the title. So the full sentence would be: "The ApiKeyValueMapUpload step allows you to prevent command execution in case the/an API key value map already exists." While the other bullet points are full sentences: "You can create new API key value map in the API portal."
> 
> Choose one style and stick to it, please. It is making it much more difficult to read otherwise.

This was not fixed!

> * If API key value map already exist, then delete it and execute the piper step again, which create new API Key value Map.

Reviewer 2: What is the difference between these two bullet points?

> * If API key value map already exist, then delete it and execute the piper step again, which create new API Key value Map.

Reviewer 2: ```suggestion
* If API key value map already exists, then delete it and execute the piper step again, which will create a new API Key value Map.
```

Author: done

> * ApiKeyValueMapUpload only support create operation, but not delete, get, update, which are supported in different piper steps.

Reviewer 2: ```suggestion
* ApiKeyValueMapUpload only supports create operation, but not delete, get, update, which are supported in different piper steps.
```
If the other piper steps exist, please reference them here.

Author: those commands not yet implemented, this line was added as per Roland's suggestion.

Reviewer: This line seems to be misunderstanding. I wouldn't expect a step named  "Upload" to _get_ or _delete_ something. Above, you already described that you can _create_ but not _overwrite_ a key. From my point of view there is no need to mention _get_ or _delete_.

Author: fixed

# ${docGenStepName}

## ${docGenDescription}

> * This step ApiKeyValueMapUpload helps you to store one or more key value pairs of data in a grouping called a map, or key value map.

Reviewer: Either "This step stores one or ..." or "ApiKeyValueMapUpload stores one or ..."

Author: fixed

* A typical flow would look something like

> You have to store API management service key in jenkins server as secret text. this is obtained from the SAP BTP sub account cockpit, under instance and subscriptions for API management under API plan. then you create new jenkins pipeline, which consumes the ApiKeyValueMapUpload piper step. you can execute the pipeline and validate step exection results as explained in the blog [Integration Suite Piper commands](https://blogs.sap.com/2022/01/05/working-with-integration-suite-piper-commands/)

Reviewer: "Get an API management service key from ..."
"Store the key ...."
"Create a Jenkins pipeline..."
"Execute...

Capital letters at the beginning of a sentence!

Author: fixed

* With the ApiKeyValueMapUpload step

You can create API key value map in the API portal.

If API key value map already exist, then command execution would fail.

If API key value map already exist, then delete it and execute the piper step again,
which create new API Key value Map.

ApiKeyValueMapUpload only supports create operation, but not either delete, get, update, which is supported in different piper steps.

> 

Reviewer: We used to pass our texts so someone who knows the SAP standards. Could you possibly ask your product manager for assistance.

Author: done

Reviewer 2: What was done?
